Complete proxy support [bsc#1043538]

flavio opened the pull request:
This pull request completes the work around handling proxy settings from velum.

It implements missing features like:

  * Handle the `proxy:systemwide` pillar that was introduced by [this](https://github.com/kubic-project/salt/pull/107) pull request inside of the salt repository.
  * Handle proxy settings inside of AutoYaST, fixing [bsc#1043538](https://bugzilla.suse.com/show_bug.cgi?id=1043538).

It also fixes previously uncovered bugs:
  * Not being able to unset certain pillars from velum (see [this](https://github.com/kubic-project/velum/commit/1e877bafc44a00b3c301e511e4b868dc5c20524a) specific commit).

Finally it changes the UI to allow an easy way to enable/disable proxy settings. See the video below:

![velum-proxy-2017-06-28_08 29 03](https://user-images.githubusercontent.com/22728/27623418-b5cd0114-5bdc-11e7-8292-4c0604828648.gif)

flavio commented:
ping @kravciak: if this PR gets merged the openQA needles will need to be updated.

MaximilianMeister commented on app/models/pillar.rb:
```diff
@@ -44,17 +44,18 @@ def value(pillar:)
+        email:            "certificate_information:subject_properties:Email",
+        country:          "certificate_information:subject_properties:C",
+        state:            "certificate_information:subject_properties:ST",
+        city:             "certificate_information:subject_properties:L",
```
JFYI there will be some collissions with https://github.com/kubic-project/velum/pull/210 - so one or the other will need a rebase after a merge

thkukuk commented on app/views/dashboard/autoyast.xml.erb:
```diff
@@ -137,6 +137,20 @@
+      <no_proxy><%= @proxy_no_proxy %></no_proxy>
+      <% end %>
+    </proxy>
+    <% end %>
```
The &lt;proxy&gt; section in the autoyast profile will not work, since this is not executed in stage1 of autoyast.
We need to replace that with a script which replaces the variables in /etc/sysconfig/proxy directly.

mssola commented:
Adding @vitoravelino to review the frontend part of this PR 😉

vitoravelino commented on app/assets/javascripts/setup.js:
```diff
@@ -0,0 +1,5 @@
+$('.btn-toggle').click(function() {
```
I would rename it to `.btn-group-toggle` or just add `.btn-group` to make it clear that `.btn-toggle` is in the same level of a button group component.

If `.btn-group-toggle` is the choice, remember to update the view template.

vitoravelino commented on app/assets/javascripts/setup.js:
```diff
@@ -0,0 +1,5 @@
+$('.btn-toggle').click(function() {
+  $(this).find('.btn').toggleClass('active');
```
Reuse `$(this)` like `var $btnGroup = $(this);`.

vitoravelino commented on app/views/setup/welcome.html.slim:
```diff
@@ -20,23 +21,43 @@ h1 Initial CaaSP Configuration
-            a data-toggle="tooltip" data-placement="left" title="Comma separated list of sites or domains that should not be accessed via the proxy server."
-              i class='glyphicon glyphicon-info-sign'
+        .input-group.btn-group.btn-toggle.pull-right data-toggle="buttons"
+          = label_tag :proxy_toggle, nil, class: "btn btn-default", "data-toggle" => "collapse", "data-target" => "#proxy-settings-panel"
```
`data: { key: value }` syntax preferred.

vitoravelino commented on app/views/setup/welcome.html.slim:
```diff
@@ -20,23 +21,43 @@ h1 Initial CaaSP Configuration
+          = label_tag :proxy_toggle, nil, class: "btn btn-default", "data-toggle" => "collapse", "data-target" => "#proxy-settings-panel"
+            = f.radio_button :enable_proxy, "enable", checked: false
+            | Enable
+          = label_tag :proxy_toggle, nil, class: "btn btn-primary active", "data-toggle" => "collapse", "data-target" => "#proxy-settings-panel"
```
Same here.

vitoravelino commented:
Rubocop is also complaining. Forgot to mention that.

vitoravelino commented:
Rebase also needed.

flavio commented:
I fixed all the rubocop issues, @vitoravelino's comments and I rebased.

Now I have to take care of @thkukuk's feedback.

vitoravelino commented on app/assets/javascripts/setup.js:
```diff
@@ -0,0 +1,6 @@
+$('.btn-group-toggle').click(function() {
+  $btnGroup = $(this);
```
Forgot `var` keyword before `$btnGroup`. This only works because we are not in `use strict` (recommended btw) mode. 

After 1.0 we will add eslint and move to ES6 (hopefully) and these catches won't need manual review.

vitoravelino replied:
If you do that, you will declare `$btnGroup` as global (attached to window) and not in the function's scope.

vitoravelino commented on app/views/setup/welcome.html.slim:
```diff
@@ -20,23 +21,43 @@ h1 Initial CaaSP Configuration
-          span class="input-group-addon"
-            a data-toggle="tooltip" data-placement="left" title="Comma separated list of sites or domains that should not be accessed via the proxy server."
-              i class='glyphicon glyphicon-info-sign'
+        .input-group.btn-group-toggle.btn-toggle.pull-right data-toggle="buttons"
```
Is the `.btn-toggle` still necessary?

vitoravelino commented on app/views/setup/welcome.html.slim:
```diff
@@ -20,23 +21,43 @@ h1 Initial CaaSP Configuration
+          = f.label :proxy_systemwide, "Use proxy systemwide"
+          p Choose whether the proxy settings should apply only to the container
+            engine or to all the processes running on the cluster nodes.
+          .input-group.btn-group-toggle.btn-toggle data-toggle="buttons"
```
Same here for `.btn-toggle`.

vitoravelino commented on app/views/setup/welcome.html.slim:
```diff
@@ -20,23 +21,43 @@ h1 Initial CaaSP Configuration
-        = f.url_field :https_proxy, value: Pillar.value(pillar: :https_proxy), class: "form-control"
+    .panel-heading.clearfix
+      | Proxy settings
       .form-group
```
Add `.pull-right` to float to the right.

flavio replied:
I tried adding that but it screwed up everything. The `pull-right` is already a couple of lines below, it's attached to the `div` that includes the labels and radio buttons.

vitoravelino replied:
You can remove `.form-group` (it will remove the margin-bottom) since the button group is not being used in a typical form.

vitoravelino commented on app/views/setup/welcome.html.slim:
```diff
@@ -20,23 +21,43 @@ h1 Initial CaaSP Configuration
-        = f.label :https_proxy, "HTTPS proxy"
-        = f.url_field :https_proxy, value: Pillar.value(pillar: :https_proxy), class: "form-control"
+    .panel-heading.clearfix
+      | Proxy settings
```
Can you add `h3.panel-title` here? It will make the panel title font work out of the box when https://github.com/kubic-project/velum/pull/213 is merged.

flavio commented:
@thkukuk : do you have any idea why AutoYaST sets also the `/root/.curlrc` file? Aren't the `/etc/sysconfig/proxy` settings enough?

I'm asking because if that's not enough we should also change our salt states to deal with `/root/.curlrc` changes too (I've already the code done locally, but I don't want to push something useless/redundant).

PI-Victor commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Bug 1043538 - AutoYaST profile has no proxy configuration](https://trello.com/c/K1VL9LTn/96-bug-1043538-autoyast-profile-has-no-proxy-configuration)

thkukuk commented:
On Thu, Jun 29, Flavio Castelli wrote:

> @thkukuk : do you have any idea why AutoYaST sets also the /root/.curlrc file?
> Aren't the /etc/sysconfig/proxy settings enough?

I couldn't find anybody who remembers why, the people who should know more
are on vacation.
But our best guess is: zypper is using libcurl. So if you run zypper from
a cronjob or su, the /etc/sysconfig/proxy variables are not parsed and set
in the environment. Which means, zypper will not use the proxy and fail.
With /root/.curlrc, libcurl will use the proxies configured there.

-- 
Thorsten Kukuk, Distinguished Engineer, Senior Architect SLES & CaaSP
SUSE LINUX GmbH, Maxfeldstr. 5, 90409 Nuernberg, Germany
GF: Felix Imendoerffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nuernberg)

flavio commented:
Thanks @thkukuk, I created https://github.com/kubic-project/salt/pull/140 to handle `/root/.curlrc` also from salt.

kiall commented on app/views/dashboard/autoyast.xml.erb:
```diff
@@ -40,6 +40,32 @@
 <% end %>
     </chroot-scripts>
+<% if @proxy_systemwide %>
+    <postpartitioning-scripts config:type="list">
```
Why postpartitioning-scripts instead of chroot-scripts? If proxies are normally done in stage 2 (post reboot), then why would we do them so early? Will yast fix the permissions on the 2 folder we create? /root needing to being go-rwx and /etc/sysconfig being go+rx

flavio replied:
I looks like the permissions are correct:
```
api:~ # ls -ld /root/
drwx------ 1 root root 82 Jun 29 08:22 /root/
api:~ # ls -ld /etc/sysconfig/
drwxr-xr-x 1 root root 532 Jun 29 08:21 /etc/sysconfig/
```

thkukuk commented:
On Thu, Jun 29, Kiall Mac Innes wrote:

> @kiall commented on this pull request.
> 
> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
> 
> In app/views/dashboard/autoyast.xml.erb:
> 
> > @@ -40,6 +40,32 @@
>        </script>
>  <% end %>
>      </chroot-scripts>
> +<% if @proxy_systemwide %>
> +    <postpartitioning-scripts config:type="list">
> 
> Why postpartitioning-scripts instead of chroot-scripts? If proxies are normally
> done in stage 2 (post reboot), then why would we do them so early? Will yast
> fix the permissions on the 2 folder we create? /root needing to being go-rwx
> and /etc/sysconfig being go+rx

I think Flavio was using postpartitioning-scripts, because it's easier
to cate the proxy file.
But correct is really to do it as chroot-scripts like the others, and
not use echo, but sed to adjust the variables in /etc/sysconfig/proxy

-- 
Thorsten Kukuk, Distinguished Engineer, Senior Architect SLES & CaaSP
SUSE LINUX GmbH, Maxfeldstr. 5, 90409 Nuernberg, Germany
GF: Felix Imendoerffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nuernberg)

flavio commented:
I used a post-partitioning script because the chroot ones are executed after the system has been installed, right before rebooting it (at least that's what I got reading [this](https://www.suse.com/documentation/sles-12/book_autoyast/data/createprofile_scripts.html) page).

Assuming a customer has a proxy configured inside of his network and doesn't have SMT (hence packages are taken from SCC), we need to set the proxy before zypper is ever invoked otherwise it won't be possible to download any package.

That was my reasoning.

thkukuk commented:
Hi,

On Thu, Jun 29, Flavio Castelli wrote:

> I used a post-partitioning script because the chroot ones are executed after
> the system has been installed, right before rebooting it (at least that's what
> I got reading this page).
> 
> Assuming a customer has a proxy configured inside of his network and doesn't
> have SMT (hence packages are taken from SCC), we need to set the proxy before
> zypper is ever invoked otherwise it won't be possible to download any package.
> 
> That was my reasoning.

zypper is not using this values from the to be installed system, but
from the inst-sys (we call the equivalent of zypper -R ...). So no need
to write this variables very early.

-- 
Thorsten Kukuk, Distinguished Engineer, Senior Architect SLES & CaaSP
SUSE LINUX GmbH, Maxfeldstr. 5, 90409 Nuernberg, Germany
GF: Felix Imendoerffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nuernberg)

flavio commented:
> zypper is not using this values from the to be installed system, but from the inst-sys (we call the
> equivalent of zypper -R ...). So no need to write this variables very early.

Good to know, I'll change the code to create a chroot script then.

vitoravelino commented on app/views/setup/welcome.html.slim:
```diff
@@ -20,23 +21,43 @@ h1 Initial CaaSP Configuration
-          span class="input-group-addon"
-            a data-toggle="tooltip" data-placement="left" title="Comma separated list of sites or domains that should not be accessed via the proxy server."
-              i class='glyphicon glyphicon-info-sign'
+        .input-group.btn-group-toggle.pull-right data-toggle="buttons"
```
Add `.btn-group-sm` to fit the same style as in #213 (screenshots update missing for this change).

vitoravelino replied:
Just realized that you are using [input group](http://getbootstrap.com/components/#input-groups) instead of [button group](http://getbootstrap.com/components/#btn-groups). Apparently there's no difference but not sure if `.btn-group-sm` will work as expect if input group.

flavio replied:
That works fine, you can definitely see the difference:

![screenshot-localhost 3000 2017-06-30 11-05-18](https://user-images.githubusercontent.com/22728/27728988-52521b02-5d84-11e7-9618-6c80c9768b83.png)

mssola commented:
There's a ruby syntax error, and rubocop is failing as well. Feel free to ping me whenever it's ready to be reviewed again 😉

flavio commented:
@mssola: I fixed the issues. Sorry, the code was broken after I rebased and I wasn't able to run the tests locally because of devenv being broken by the recent changes.

flavio commented:
I just squashed the commits together to remove some clutter. Forcefully merging.
